In [1]:
import pandas as pd

In [4]:
df = pd.read_excel("../data/raw/online_retail.xlsx")

In [9]:
df.head(5)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [10]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [11]:
df.shape

(541909, 8)

In [16]:
df["InvoiceDate"].min()

Timestamp('2010-12-01 08:26:00')

In [17]:
df["InvoiceDate"].max()

Timestamp('2011-12-09 12:50:00')

In [18]:
df.isna().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [20]:
df.duplicated().sum()

np.int64(5268)

In [21]:
df[df.duplicated()].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
587,536412,22273,FELTCRAFT DOLL MOLLY,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
589,536412,22749,FELTCRAFT PRINCESS CHARLOTTE DOLL,1,2010-12-01 11:49:00,3.75,17920.0,United Kingdom
594,536412,22141,CHRISTMAS CRAFT TREE TOP ANGEL,1,2010-12-01 11:49:00,2.10,17920.0,United Kingdom
598,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920.0,United Kingdom
600,536412,22569,FELTCRAFT CUSHION BUTTERFLY,2,2010-12-01 11:49:00,3.75,17920.0,United Kingdom


In [24]:
df["InvoiceNo"].value_counts().head(10)

InvoiceNo
573585    1114
581219     749
581492     731
580729     721
558475     705
579777     687
581217     676
537434     675
580730     662
538071     652
Name: count, dtype: int64

In [25]:
df.value_counts().head(10)

InvoiceNo  StockCode  Description                          Quantity  InvoiceDate          UnitPrice  CustomerID  Country       
555524     22698      PINK REGENCY TEACUP AND SAUCER       1         2011-06-05 11:37:00  2.95       16923.0     United Kingdom    20
           22697      GREEN REGENCY TEACUP AND SAUCER      1         2011-06-05 11:37:00  2.95       16923.0     United Kingdom    12
572861     22775      PURPLE DRAWERKNOB ACRYLIC EDWARDIAN  12        2011-10-26 12:46:00  1.25       14102.0     United Kingdom     8
538514     21756      BATH BUILDING BLOCK WORD             1         2010-12-12 14:27:00  5.95       15044.0     United Kingdom     6
540524     21756      BATH BUILDING BLOCK WORD             1         2011-01-09 12:53:00  5.95       16735.0     United Kingdom     6
541266     21755      LOVE BUILDING BLOCK WORD             1         2011-01-16 16:25:00  5.95       15673.0     United Kingdom     6
           21754      HOME BUILDING BLOCK WORD             1        

### Duplicate Records

The dataset contains 5,268 rows identified as exact duplicates
across all columns.

Because the dataset is at invoice-line grain, retaining exact
duplicate rows could artificially inflate product demand and revenue.

We will therefore remove exact duplicate records before constructing
the demand dataset.

In [30]:
df[df["Quantity"] < 0].shape[0]

10624

In [31]:
df[df["Quantity"] < 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [33]:
df["InvoiceNo"].dtype

dtype('O')

In [34]:
df["InvoiceNo"].map(type).value_counts()

InvoiceNo
<class 'int'>    532618
<class 'str'>      9291
Name: count, dtype: int64

In [35]:
df[df["InvoiceNo"].map(type) == str]["InvoiceNo"].value_counts().head(20)

InvoiceNo
C570867    101
C560540     57
C548460     45
C560855     41
C538341     39
C570828     39
C557663     38
C569985     36
C542426     35
C559939     32
C560409     31
C562375     31
C569655     31
C570221     30
C547725     27
C538350     26
C568879     26
C569989     26
C570290     25
C565848     24
Name: count, dtype: int64

In [36]:
df[df["InvoiceNo"].map(type) == str].head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [37]:
df[df["InvoiceNo"].map(type) == str]["Quantity"].describe()

count     9291.000000
mean       -29.875256
std       1145.602081
min     -80995.000000
25%         -6.000000
50%         -2.000000
75%         -1.000000
max          1.000000
Name: Quantity, dtype: float64

In [38]:
df[df["InvoiceNo"].map(type) == str]["Quantity"].value_counts().head(20)

Quantity
-1      4105
-2      1337
-3       581
-12      546
-6       482
-4       467
-24      244
-5       189
-10      168
-8       148
-48       88
-7        82
-36       59
-9        55
-100      50
-20       47
-11       47
-16       45
-25       43
-18       35
Name: count, dtype: int64

In [39]:
df[(df["InvoiceNo"].map(type) == str) & (df["Quantity"] > 0)]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299982,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [40]:
df[df["Quantity"] == df["Quantity"].min()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
540422,C581484,23843,"PAPER CRAFT , LITTLE BIRDIE",-80995,2011-12-09 09:27:00,2.08,16446.0,United Kingdom


In [43]:
df["InvoicePrefix"] = df["InvoiceNo"].astype(str).str[0]

df["InvoicePrefix"].value_counts()

InvoicePrefix
5    532618
C      9288
A         3
Name: count, dtype: int64

In [44]:
df.groupby("InvoicePrefix").agg(
    rows=("InvoiceNo", "size"),
    total_quantity=("Quantity", "sum"),
    total_revenue=("UnitPrice", "sum"),
).sort_values("rows", ascending=False)

,rows,total_quantity,total_revenue
InvoicePrefix,,,
5,532618,5454021,2060385.714
C,9288,-277574,449480.320
A,3,3,-11062.060


In [42]:
df[df["InvoicePrefix"] == "A"]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,InvoicePrefix
299982,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom,A
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom,A
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom,A


In [45]:
df[
    (df["InvoiceNo"].map(type) == str)
    & (~df["InvoiceNo"].astype(str).str.startswith(("C", "A")))
][["InvoiceNo", "StockCode", "Description", "Quantity", "UnitPrice"]]

,InvoiceNo,StockCode,Description,Quantity,UnitPrice


### Transaction types
C... InvoiceNo values are associated with cancellation/return transactions, while A... records are special bad-debt adjustments. These should not be treated as ordinary product demand.

In [46]:
df["StockCode"].isna().sum()

np.int64(0)

In [47]:
df.groupby("InvoiceNo")["StockCode"].nunique().describe()

count    25900.000000
mean        20.510618
std         42.500488
min          1.000000
25%          2.000000
50%         10.000000
75%         23.000000
max       1110.000000
Name: StockCode, dtype: float64

In [48]:
df.groupby("InvoiceNo")["StockCode"].apply(lambda x: x.isna().any()).sum()

np.int64(0)

In [49]:
df["UnitPrice"].describe()

count    541909.000000
mean          4.611114
std          96.759853
min      -11062.060000
25%           1.250000
50%           2.080000
75%           4.130000
max       38970.000000
Name: UnitPrice, dtype: float64

In [50]:
(df["UnitPrice"] <= 0).sum()

np.int64(2517)

In [56]:
df[df["UnitPrice"] <= 0][
    ["InvoiceNo", "StockCode", "Description", "Quantity", "UnitPrice"]
].head(20)

,InvoiceNo,StockCode,Description,Quantity,UnitPrice
622,536414,22139,NaN,56,0.0
1970,536545,21134,NaN,1,0.0
1971,536546,22145,NaN,1,0.0
1972,536547,37509,NaN,1,0.0
1987,536549,85226A,NaN,1,0.0
1988,536550,85044,NaN,1,0.0
2024,536552,20950,NaN,1,0.0
2025,536553,37461,NaN,3,0.0
2026,536554,84670,NaN,23,0.0
2406,536589,21777,NaN,-10,0.0


### UnitPrice anomalies
UnitPrice contains 2,517 records with values ≤ 0. The dataset also contains extreme positive and negative prices. These records include special transactions such as zero-price lines and non-standard entries, so they require investigation before deciding whether they should be excluded from the demand/revenue dataset.

In [57]:
df[df["UnitPrice"] <= 0]["Description"].value_counts().head(20)

Description
check                              159
?                                   47
damages                             45
damaged                             43
found                               25
sold as set on dotcom               20
adjustment                          16
Damaged                             14
FRENCH BLUE METAL DOOR SIGN 1        9
thrown away                          9
Unsaleable, destroyed.               9
amazon                               8
FRENCH BLUE METAL DOOR SIGN 8        8
Found                                8
FRENCH BLUE METAL DOOR SIGN 4        7
FRENCH BLUE METAL DOOR SIGN No       7
OWL DOORSTOP                         7
FRENCH BLUE METAL DOOR SIGN 3        7
RECIPE BOX PANTRY YELLOW DESIGN      7
??                                   7
Name: count, dtype: int64

In [58]:
df[df["UnitPrice"] <= 0]["StockCode"].value_counts().head(20)

StockCode
23084     16
35965     13
20713     11
22501     10
22676      9
22627      9
22084      9
22683      8
21116      8
46000S     8
82583      8
22734      7
22679      7
22686      7
22625      7
79321      7
22678      7
22666      7
23348      7
21830      7
Name: count, dtype: int64

In [59]:
df[df["UnitPrice"] <= 0]["InvoiceNo"].astype(str).str[0].value_counts()

InvoiceNo
5    2515
A       2
Name: count, dtype: int64

### Non-positive prices

There are 2,517 transactions with UnitPrice <= 0. These include descriptions indicating operational adjustments, damaged/unsaleable goods, checks, and other non-standard transactions. Therefore, UnitPrice <= 0 records should not automatically be treated as normal product sales and require an explicit filtering/business rule when constructing the demand dataset.

In [61]:
df[df["UnitPrice"] == df["UnitPrice"].min()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,InvoicePrefix
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom,A
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom,A


In [62]:
df[df["UnitPrice"] == df["UnitPrice"].max()]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,InvoicePrefix
222681,C556445,M,Manual,-1,2011-06-10 15:31:00,38970.0,15098.0,United Kingdom,C


### Extreme prices

The minimum and maximum UnitPrice values are associated with special transactions rather than ordinary product sales: A... bad-debt adjustments and a C... manual cancellation record. Therefore, extreme prices should not be interpreted as genuine product prices when analyzing demand or revenue.

In [63]:
df["StockCode"].value_counts().head(30)

StockCode
85123A    2313
22423     2203
85099B    2159
47566     1727
20725     1639
84879     1502
22720     1477
22197     1476
21212     1385
20727     1350
22383     1348
22457     1280
23203     1267
POST      1256
22386     1251
22469     1239
22960     1229
21931     1214
22086     1210
22411     1202
20728     1197
22382     1192
22961     1182
22666     1180
23298     1179
22699     1138
22384     1137
23209     1135
82482     1129
22993     1111
Name: count, dtype: int64

In [64]:
df["StockCode"].nunique()

4070

In [66]:
df["StockCode"].map(type).value_counts()

StockCode
<class 'int'>    487036
<class 'str'>     54873
Name: count, dtype: int64

In [67]:
df[df["StockCode"].map(type) == int]["StockCode"].value_counts().head(20)

StockCode
22423    2203
47566    1727
20725    1639
84879    1502
22720    1477
22197    1476
21212    1385
20727    1350
22383    1348
22457    1280
23203    1267
22386    1251
22469    1239
22960    1229
21931    1214
22086    1210
22411    1202
20728    1197
22382    1192
22961    1182
Name: count, dtype: int64

In [68]:
df[df["StockCode"].map(type) == str]["StockCode"].value_counts().head(30)

StockCode
85123A     2313
85099B     2159
POST       1256
85099C      960
82494L      943
85099F      842
DOT         710
M           571
84970S      546
84596B      541
84997D      481
84029G      474
15056N      467
84970L      460
84029E      452
84596F      452
47591D      446
85049E      423
47590B      402
47590A      394
47566B      379
84997B      377
85014B      355
84030E      346
84536A      346
84997C      339
85049A      332
15056BL     326
84032A      318
84406B      296
Name: count, dtype: int64